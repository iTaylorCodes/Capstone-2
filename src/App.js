import { useState, useEffect } from "react";
import jwt from "jsonwebtoken";
import { NeighborhoodApi } from "./api/api";
import NavBar from "./routes-nav/NavBar";
import Router from "./routes-nav/Router";
import useLocalStorage from "./hooks/useLocalStorage";
import UserContext from "./auth/UserContext";

// Key name for storing token in localStorage
export const LOCAL_STORAGE_TOKEN = "neighborhood-token";

function App() {
  const [currentUser, setCurrentUser] = useState(null);
  const [token, setToken] = useLocalStorage(LOCAL_STORAGE_TOKEN);

  // Gets the current user if token in localStorage
  useEffect(() => {
    const getCurrentUser = async () => {
      if (token) {
        try {
          let { username } = jwt.decode(token);
          NeighborhoodApi.token = token;
          let currentUser = await NeighborhoodApi.getCurrentUser(username);
          setCurrentUser(currentUser);
        } catch (e) {
          console.error("Error Losing user info", e);
          setCurrentUser(null);
        }
      }
    };
    getCurrentUser();
  }, [token]);

  // Handles signing up
  const signup = async (signupData) => {
    try {
      let token = await NeighborhoodApi.signup(signupData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("Signup attempt failed", errors);
      return { success: false, errors };
    }
  };

  // Handles logging out
  const logout = () => {
    setCurrentUser(null);
    setToken(null);
  };

  // Handles login
  const login = async (loginData) => {
    try {
      let token = await NeighborhoodApi.login(loginData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("Login attempt failed", errors);
      return { success: false, errors };
    }
  };

  return (
    <UserContext.Provider value={{ currentUser, setCurrentUser }}>
      <div className="App">
        <NavBar logout={logout} />
        <div>
          <Router signup={signup} login={login} />
        </div>
      </div>
    </UserContext.Provider>
  );
}

export default App;
