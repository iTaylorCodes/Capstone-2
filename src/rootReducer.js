import { v4 as uuidv4 } from "uuid";

const INITIAL_STATE = {
  accountWasDeleted: false,
  recentSearches: [],
};

const rootReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case "ACCOUNT_DELETED":
      return {
        ...state,
        accountWasDeleted: true,
      };

    case "RESET_ACCOUNT_DELETED":
      return {
        ...state,
        accountWasDeleted: false,
      };

    case "NEW_SEARCH":
      if (state.recentSearches.length >= 5) {
        state.recentSearches.shift();
      }

      return {
        ...state,
        recentSearches: [
          ...state.recentSearches,
          {
            city: action.searchTerm,
            image: action.image,
            scores: action.scores,
            id: uuidv4(),
          },
        ],
      };

    default:
      return state;
  }
};

export default rootReducer;
