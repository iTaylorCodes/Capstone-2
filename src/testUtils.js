import UserContext from "./auth/UserContext";

const testUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
};

const UserProvider = ({ children, currentUser = testUser }) => (
  <UserContext.Provider value={{ currentUser }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };
