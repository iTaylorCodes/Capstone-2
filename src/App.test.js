import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import App from "./App";
import { UserProvider } from "./testUtils";
import { Provider } from "react-redux";
import { createStore } from "redux";
import rootReducer from "./rootReducer";

const store = createStore(rootReducer);

it("renders without crashing", function () {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <UserProvider>
          <App />
        </UserProvider>
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <UserProvider>
          <App />
        </UserProvider>
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
