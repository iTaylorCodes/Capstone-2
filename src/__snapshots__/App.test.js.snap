// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="NavBar navbar navbar-expand-lg navbar-light bg-light"
    >
      <div
        class="container-fluid"
      >
        <a
          aria-current="page"
          class="navbar-brand text-primary fs-4 fw-bold"
          href="/"
        >
          <img
            alt="Logo"
            class="me-2"
            src="Neighborhoodz-NavBar-Logo.png"
            style="width: 35px; height: 35px;"
          />
          Neighborhoodz
        </a>
        <button
          aria-controls="navbarTogglerDemo02"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-bs-target="#navbarTogglerDemo02"
          data-bs-toggle="collapse"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
          id="navbarTogglerDemo02"
        >
          <ul
            class="navbar-nav me-auto mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                href="/signup"
              >
                Signup
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                href="/login"
              >
                Login
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <div>
      <div
        class="text-center home-hero"
      >
        <div
          class="HomeContent mt-3"
        >
          <div
            class="Home-heading mb-3"
          >
            <h1
              class="text-primary mb-3"
            >
              <img
                alt="Logo"
                class="me-2 Home-logo"
                src="Neighborhoodz-Logo.png"
              />
              Neighborhoodz
            </h1>
            <h3>
              Find your new Neighborhood!
            </h3>
          </div>
          <div
            class="fw-bold"
            id="alert-container"
          />
          <div
            class="SearchBar d-flex justify-content-center"
          >
            <div
              class="mb-3"
            >
              <form
                class="input-group"
              >
                <div
                  class="form-floating"
                >
                  <input
                    class="form-control ps-5 pe-5"
                    id="city-search"
                    name="searchTerm"
                    placeholder="Search City, State:"
                    value=""
                  />
                  <label
                    for="city-search"
                  >
                    Search City, State:
                  </label>
                </div>
                <button
                  class="btn btn-primary"
                  data-testid="search-button"
                  type="submit"
                >
                  <i
                    class="fa fa-search"
                  />
                </button>
              </form>
            </div>
          </div>
          <div
            class="SearchesCarousel container d-flex justify-content-center"
          >
            <div
              class="col-12 col-md-8 col-lg-6 col-xl-5 col-xxl-4 bg-light rounded text-center p-1"
            >
              <div
                class="row mb-1"
              >
                <div
                  class="col-11 d-flex justify-content-center"
                >
                  <h3>
                    Recent Searches
                  </h3>
                </div>
              </div>
              <p>
                Your most recent searches will be shown here.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
